File: src/project_manager_flutter/lib/features/projects/repository/project_repository.dart
import 'package:dio/dio.dart';
import '../../../core/constants/api_constants.dart';
import '../../../core/network/api_exceptions.dart';
import '../models/project.dart';
import '../models/project_detail.dart';
import '../models/create_project_request.dart';
import '../models/update_project_request.dart';

class ProjectRepository {
  final Dio _dio;

  ProjectRepository(this._dio);

  Future<List<Project>> listProjects({String? clientId}) async {
    try {
      final response = await _dio.get(
        ApiConstants.projects,
        queryParameters: {if (clientId != null) 'clientId': clientId},
      );
      return (response.data as List).map((e) => Project.fromJson(e)).toList();
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }

  Future<ProjectDetail> getProject(String id) async {
    try {
      final response = await _dio.get(ApiConstants.projectById(id));
      return ProjectDetail.fromJson(response.data);
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }

  Future<void> createProject(CreateProjectRequest request) async {
    try {
      await _dio.post(ApiConstants.projects, data: request.toJson());
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }

  Future<void> updateProject(String id, UpdateProjectRequest request) async {
    try {
      await _dio.put(ApiConstants.projectById(id), data: request.toJson());
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }

  Future<void> deleteProject(String id) async {
    try {
      await _dio.delete(ApiConstants.projectById(id));
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }
}

File: src/project-manager-web/src/features/projects/repository/project-repository.ts
import { apiClient } from "@/core/api/client";
import { ApiEndpoints } from "@/core/api/constants";
import { ApiException } from "@/core/api/exceptions";
import type { Project, ProjectDetail, CreateProjectRequest, UpdateProjectRequest } from "../models/types";
import { AxiosError } from "axios";

function handleError(error: unknown): never {
  if (error instanceof AxiosError) {
    throw ApiException.fromResponse(error.response?.status ?? 500, error.response?.data);
  }
  throw error;
}

export const projectRepository = {
  async list(clientId?: string): Promise<Project[]> {
    try {
      const response = await apiClient.get<Project[]>(ApiEndpoints.projects, {
        params: clientId ? { clientId } : undefined,
      });
      return response.data;
    } catch (error) { handleError(error); }
  },

  async get(id: string): Promise<ProjectDetail> {
    try {
      const response = await apiClient.get<ProjectDetail>(ApiEndpoints.projectById(id));
      return response.data;
    } catch (error) { handleError(error); }
  },

  async create(request: CreateProjectRequest): Promise<void> {
    try {
      await apiClient.post(ApiEndpoints.projects, request);
    } catch (error) { handleError(error); }
  },

  async update(id: string, request: UpdateProjectRequest): Promise<void> {
    try {
      await apiClient.put(ApiEndpoints.projectById(id), request);
    } catch (error) { handleError(error); }
  },

  async delete(id: string): Promise<void> {
    try {
      await apiClient.delete(ApiEndpoints.projectById(id));
    } catch (error) { handleError(error); }
  },
};
